# Bài thực hành 2
### Môn: Linear Algebra

In [1]:
import numpy as np

### Bài tập 1
Cho ma trận ${ A}=\left[\begin{matrix}
0 & 0 & 0\\
1 & 0 & 0\\
0 & 1 & 0
\end{matrix}\right]\ \ ;\ \ { B}=\left[\begin{matrix}
4 & -2 & 1\\
3 & 1 & -1\\
-1 & 5 & 0
\end{matrix}\right]$

* Chứng tỏ rằng ${A}^3=0$. Tính $({I}-{A})({I}+{A}+{A}^2)$.
* Tạo ma trận khối $C=\left[\begin{matrix}
A & I\\
-I & B
\end{matrix}\right]$. Tính định thức và hạng của $C$.

In [2]:
A = np.array([0,0,0,1,0,0,0,1,0]).reshape(3,3)
B = np.array([4,-2,1,3,1,-1,-1,5,0]).reshape(3,3)

# Chứng tỏ A^3=0
print("A^3==0:", np.allclose(A@A@A, np.zeros((3,3))))

# Tính giá trị biểu thức
I = np.identity(3)
print("Giá trị biểu thức:")
print(np.dot(I-A, I+A+np.linalg.matrix_power(A,2)))

A^3==0: True
Giá trị biểu thức:
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [3]:
# Tạo ma trận khối
C = np.vstack([np.hstack([A,I]), np.hstack([-1*I,B])])
print("Ma trận khối: C=")
print(C)

# Tính định thức và hạng
print("det(C)=", np.linalg.det(C))
print("rank(C)=", np.linalg.matrix_rank(C))

Ma trận khối: C=
[[ 0.  0.  0.  1.  0.  0.]
 [ 1.  0.  0.  0.  1.  0.]
 [ 0.  1.  0.  0.  0.  1.]
 [-1. -0. -0.  4. -2.  1.]
 [-0. -1. -0.  3.  1. -1.]
 [-0. -0. -1. -1.  5.  0.]]
det(C)= -0.9999999999999998
rank(C)= 6


### Bài tập 2
Lũy thừa ma trận có thể được sử dụng như một công cụ để phân tích trạng thái tới hạn của một số hệ động lực. Cho hai ma trận sau đây:

$${\bf A}=\left[\begin{matrix}
0.4 & 0.2 & 0.3\\
0.3 & 0.6 & 0.3\\
0.3 & 0.2 & 0.4
\end{matrix}\right],\,{\bf B}=\left[\begin{matrix}
0 & 0.2 & 0.3\\
0.1 & 0.6 & 0.3\\
0.9 & 0.2 & 0.4
\end{matrix}\right]$$

Hãy kiểm tra xem điều gì sẽ xảy ra với ${\bf A}^k$ và ${\bf B}^k$ khi $k$ nhận giá trị lớn.

In [4]:
A = np.array([0.4,0.2,0.3,0.3,0.6,0.3,0.3,0.2,0.4]).reshape(3,3)
B = np.array([0,0.2,0.3,0.1,0.6,0.3,0.9,0.2,0.4]).reshape(3,3)
print("Lũy thừa ma trận A:\n")
n = 100 # Số mũ
for i in range(n):
    if i>n-5:
        print(np.linalg.matrix_power(A,i+1),'\n')

print("Lũy thừa ma trận B:\n")
for i in range(n):
    if i>n-5:
        print(np.linalg.matrix_power(B,i+1),'\n')

Lũy thừa ma trận A:

[[0.28571429 0.28571429 0.28571429]
 [0.42857143 0.42857143 0.42857143]
 [0.28571429 0.28571429 0.28571429]] 

[[0.28571429 0.28571429 0.28571429]
 [0.42857143 0.42857143 0.42857143]
 [0.28571429 0.28571429 0.28571429]] 

[[0.28571429 0.28571429 0.28571429]
 [0.42857143 0.42857143 0.42857143]
 [0.28571429 0.28571429 0.28571429]] 

[[0.28571429 0.28571429 0.28571429]
 [0.42857143 0.42857143 0.42857143]
 [0.28571429 0.28571429 0.28571429]] 

Lũy thừa ma trận B:

[[0.20224719 0.20224719 0.20224719]
 [0.37078652 0.37078652 0.37078652]
 [0.42696629 0.42696629 0.42696629]] 

[[0.20224719 0.20224719 0.20224719]
 [0.37078652 0.37078652 0.37078652]
 [0.42696629 0.42696629 0.42696629]] 

[[0.20224719 0.20224719 0.20224719]
 [0.37078652 0.37078652 0.37078652]
 [0.42696629 0.42696629 0.42696629]] 

[[0.20224719 0.20224719 0.20224719]
 [0.37078652 0.37078652 0.37078652]
 [0.42696629 0.42696629 0.42696629]] 



In [5]:
# Giá trị riêng và véc tơ riêng của A
np.linalg.eig(A)

(array([1. , 0.1, 0.3]),
 array([[-4.85071250e-01, -7.07106781e-01,  4.08248290e-01],
        [-7.27606875e-01,  2.38524432e-16, -8.16496581e-01],
        [-4.85071250e-01,  7.07106781e-01,  4.08248290e-01]]))

In [6]:
# Giá trị riêng và véc tơ riêng của B
np.linalg.eig(B)

(array([-0.33166248,  0.33166248,  1.        ]),
 array([[ 0.59404463, -0.12024109, -0.33675758],
        [ 0.18808926,  0.75951783, -0.6173889 ],
        [-0.78213388, -0.63927674, -0.71093268]]))

### Nhận xét
$A^k$ và $B^k$ hội tụ về các ma trận có dạng:

$$\left[\begin{matrix}
0.2857 & 0.2857 & 0.2857\\
0.4285 & 0.4285 & 0.4285\\
0.2857 & 0.2857 & 0.2857
\end{matrix}\right]\hspace{2cm}
\left[\begin{matrix}
0.2022 & 0.2022 & 0.2022\\
0.3707 & 0.3707 & 0.3707\\
0.4269 & 0.4269 & 0.4269
\end{matrix}\right]$$

khi $k\rightarrow\infty$. Thực tế, hai ma trận $A$ và $B$ có một trị riêng bằng 1, và hai trị riêng có độ lớn nhỏ hơn 1. Lũy thừa của các trị riêng có độ lớn nhỏ hơn 1 sẽ tiến dần đến 0.

### Bài 3
Phân tích ma trận đánh giá
$${\bf X}=\left[\begin{matrix}
3 & 3 & 0 & 4 & 4\\
0 & 5 & 2 & 0 & 1\\
2 & 0 & 1 & 2 & 5 \\
0 & 3 & 2 & 5 & 0 \\
2 & 5 & 0 & 3 & 2
\end{matrix}\right]
$$
thành tích $\bf W\,H^T$, với ${\bf W}\in\mathbb{R}^{5\times 3}$, ${\bf H}\in\mathbb{R}^{5\times 3}$. Sử dụng hàm mất mát dạng:
$$\mathcal{L}({\bf W},{\bf H})=\sum_{(u,i):x_{ui}\neq 0}\left(x_{ui}-\sum^3_{j=1}w_{uj}h_{ij}\right)^2$$

In [7]:
"""
@INPUT:
    A     : a matrix to be factorized, dimension M x N
    W     : an initial matrix of dimension M x K
    H     : an initial matrix of dimension N x K
    K     : the number of latent features
    steps : the maximum number of steps to perform the optimisation
    beta  : the learning rate
@OUTPUT:
    the final matrices W and H
"""
def matrix_factor(A, W, H, K, steps=5000, beta=0.0002):
    H = H.T
    for step in range(steps):
        for i in range(len(A)):
            for j in range(len(A[i])):
                if A[i][j] > 0:
                    eij = A[i][j] - np.dot(W[i,:],H[:,j])
                    for k in range(K):
                        W[i][k] = W[i][k] + beta*(2*eij*H[k][j])
                        H[k][j] = H[k][j] + beta*(2*eij*W[i][k])
        e = 0
        for i in range(len(A)):
            for j in range(len(A[i])):
                if A[i][j] > 0:
                    e = e + pow(A[i][j]-np.dot(W[i,:],H[:,j]), 2)
        if e < 0.0001:
            break
    return W, H.T

In [8]:
# Khởi tạo ma trận
A = np.array([[3,3,0,4,4],
              [0,5,2,0,1],
              [2,0,1,2,5],
              [0,3,2,5,0],
              [2,5,0,3,2]])

M = len(A)
N = len(A[0])
K = 3

W = np.random.rand(M,K)
H = np.random.rand(N,K)

# Phân tích ma trận
nW, nH = matrix_factor(A, W, H, K)
print(nW)
print(nH)
np.dot(nW,nH.T)

[[ 1.44371222  0.91481176  0.88126081]
 [ 0.01805909  1.9757258   0.54572117]
 [ 2.06275049  0.36381973 -0.05872752]
 [ 0.91878058  0.89542962  1.57178731]
 [ 0.51445611  1.92454888  0.04241083]]
[[0.91289898 0.84721535 0.73288541]
 [0.38913493 2.46231318 0.26939481]
 [0.34028997 0.80950886 0.65278059]
 [0.78128475 1.30529489 1.96426729]
 [2.34560862 0.37784288 0.39884399]]


array([[2.73886917, 3.05175901, 1.80709896, 4.05308125, 4.08352451],
       [2.09030244, 5.01889756, 1.96174905, 3.66495633, 1.00653109],
       [2.14827594, 1.6827055 , 0.95811242, 1.971131  , 4.95244891],
       [2.74931557, 2.98578913, 2.06354227, 4.97403937, 3.12032928],
       [2.1312361 , 4.95046019, 1.76068859, 2.99734675, 1.95080508]])